### ISSUE TITLE
improve evm_snaphost / evm_revert behavior

### ISSUE THREAD

**Issue by wighawag** [nomiclabs/buidler#659]
current evm_snaphost / evm_revert is not intuitive and lead to optimization loss when used certain way

let say I run the following

1. deploy contracts
2. evm_snapshot => **S1** // save deployments to not need to reexecute
3. run test 1
4. evm_revert  S1 // revert back to **S1** to execute test 2 on clean deployment
5. evm_snapshot => S2 // necessary because evm_revert delete its own snapshot
6. run test 2
7. evm_revert  S2 // revert back to **S1** (now saved as S2)  to execute test 3 on clean deployment
8. evm_snapshot => S3 // necessary because evm_revert delete its own snapshot
9. extra deployments for test 3 involving change to deployed contracts
10. evm_snapshot => **S4** // save such new deployments
11. run test 3
12. evm_revert  S4 // revert back to **S4** to run test 4 in the same environment as test 3
13. evm_snapshot => S5 // necessary because evm_revert delete its own snapshot
14. run test 4
15. evm_revert  S3 // revert back to **S1** (now saved as S3)  to execute test 5 on clean deployment
16. evm_snapshot => S6 // necessary because evm_revert delete its own snapshot
17. extra deployments for test 5 involving change to deployed contracts
18. evm_snapshot => **S7** // save such new deployments
19. run test 5
20. evm_revert  S7 // revert back to **S7** to run test 6 in the same environment as test 5
21. evm_snapshot => S8 // necessary because evm_revert delete its own snapshot
22. run test 6
23. evm_revert S5 // revert to **S4** (now saved as S5) to go back to test 3 environment  // THIS FAIL SILENTLY
24. evm_snapshot => S6
25. run test 7 // want to execute test in the same environment as test 3

The first issue is that `evm_revert` delete its own snapshot so I need to create a new one. This is not a big deal though
But the problem come when there is multiple interleaved snapshot

step 23 fails silently. this is because when `evm_revert` is executed in step 15 to go back to clean deployments, it delete snapshot S4 (now saved as S5) so while I can simply reexcute the new deployment as I did in step 9 I am missing the evm_snapshot optimization.

It would be great if snapshot are saved in a temporary fork, so they can be reused even if older snapshot where used.

In any case, when requesting a snapshot that has been deleted, the rpc method should return an error, instead of failing silently

**Comment by alcuadrado** [comment]
This is really interesting @wighawag.

Right now, the `evm_revert` method works exactly like in Ganache. I believe it was designed in this way because Truffle does a snapshot and revert on every `contract(..)` test function. 

My worry about this is that it would lead to leaking a lot of memory, and it may also be slower.

**Comment by wighawag** [comment]
yes maybe it could be new rpc method, if perfect ganache compatibility is required with existing rpc

Regarding memory leak, it would only happen if it was misused : you normally snapshot if you intend to revert.
the memory usage could simply be part of the things to know when using that feature.

we could also add `evm_delete_snapshot` or something like that if managing memory is required.

**Comment by fvictorio** [comment]
> yes maybe it could be new rpc method, if perfect ganache compatibility is required with existing rpc

Could this be done by adding an optional param to the method, and defaulting to the current behavior? I don't know if there are methods with optional params right now though.

**Comment by alcuadrado** [comment]
Nice idea. There's plenty of methods with optional params.

**Comment by vnavascues** [comment]
I'd totally support the `evm_delete_snapshot`. I've experienced the same behaviour in some of my tests that have inner contexts where I wanted to add on top of the "after deployment" snapshot, and it was overwritten (therefore outside that context the snapshot was polluted).
